### PR TITLE
Transfer OpenVpn config from Agent to Client

### PR DESCRIPTION
**Pull request by dzeckelev**
- Create OpenVPN ClientAfterEndpointMsgSOMCGet specific function, that prepares all Client OpenVPN config files
- Allow parsed OpenVPN config to be pushed to dappctrl automatically once.

Resolves BV-236
Resolves BV-244 (Finish)

### CODE REVIEW

File: ept/config/client.go [hunk 1]
```diff
@@ -50,62 +52,119 @@ var defaultConfig = &CConf{
...
+	Cipher        string `json:"cipher"`
 	ServerAddress string
-	Port          string
-	PingRestart   string
-	Ping          string
-	ConnectRetry  string
-	Ca            string
-	CompLZO       string
+	Port          string `json:"port"`
+	PingRestart   string `json:"ping-restart"`
```
[the remark this answers is not in the document]
Comment (dzeckelev) [reply ept/config/client.go]: This was done deliberately. Json field names correspond to the names in the OpenVpn configuration

[the remark this answers is not in the document]
Comment (ababo) [reply ept/config/client.go]: Why do we need to preserve the original names? Don't see any sense.

[the remark this answers is not in the document]
Comment (dzeckelev) [reply ept/config/client.go]: This simplifies the flow. We do not control the names of the fields that the server sends us.

File: svc/dappvpn/misc.go [hunk 2]
```diff
@@ -66,3 +72,7 @@ func getCreds() (string, string) {
 
 	return user, pass
 }
+
+func writeConfig(name string, cfg *config) error {
```
[the remark this answers is not in the document]
Comment (dzeckelev) [reply svc/dappvpn/misc.go]: this function is needed here, it changes the `dappvpn.config.json` file

File: svc/dappvpn/misc.go [hunk 3]
```diff
@@ -7,9 +7,15 @@ import (
...
 	"path/filepath"
+
+	"github.com/privatix/dappctrl/util"
 )
 
-const chanPerm = 0644
+const (
+	chanPerm = 0644
+
+	jsonIdent = "    "
```
[the remark this answers is not in the document]
Comment (dzeckelev) [reply svc/dappvpn/misc.go]: this is needed for `writeConfig` function, which should be here

File: ept/config/client.go [hunk 4]
```diff
@@ -3,109 +3,164 @@ package config
...
-// ClientConfig returns config object
-func ClientConfig(srvAddr, srvPort string,
-	additionalParams []byte) (*CConf, error) {
-	if !isHost(srvAddr) || !util.IsNetPort(srvPort) {
-		return nil, ErrInput
+// ConfDeployer is needed to store information about the directory in which
+// the directories with configuration files are stored.
+type ConfDeployer struct {
+	rootDir string
+}
```
Comment (furkhat) [inline ept/config/client.go]: I don't see usages of this type outside of `config` package. Let's not export types we don't need.

File: ept/config/client_test.go [hunk 5]
```diff
@@ -47,18 +58,47 @@ func TestGetText(t *testing.T) {
...
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if isNotExist(filepath.Join(end, clientConfName)) ||
+			isNotExist(filepath.Join(end, clientAccessName)) {
+			t.Fatal(errDeployConfig)
+		}
+	}
 }
```
Comment (furkhat) [inline ept/config/client_test.go]: How do you test the output of config generation?

File: ept/config/misc.go [hunk 6]
```diff
@@ -104,3 +105,14 @@ func fillCa(results map[string]string, filePath string) error {
 
 	return nil
 }
+
+func createPath(target string) error {
+	return os.MkdirAll(target, pathPerm)
+}
+
+func isNotExist(object string) bool {
```
Comment (furkhat) [inline ept/config/misc.go]: `func notExist(location string) bool`

File: svc/dappvpn/pusher/pusher.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	"context"
+
+	c "github.com/privatix/dappctrl/ept/config"
+	"github.com/privatix/dappctrl/util"
+	"github.com/privatix/dappctrl/util/srv"
+)
+
+// Config for pushing OpenVpn configuration.
+type Config struct {
+	ExportConfigKeys []string
```
Comment (furkhat) [inline svc/dappvpn/pusher/pusher.go]: Please add comments describing the use of `ExportConfigKeys`

File: ept/config/client.go (hunk 4 above)
Comment (dzeckelev) [reply ept/config/client.go]: `rootDir`  does not change during the execution of the program.
Why not save `rootDir`  field once.
I see no reason to use global variables for this.

File: dappctrl-test.config.json [hunk 8]
```diff
@@ -122,5 +126,11 @@
 
     "VPNMonitorTest": {
         "ServerStartupDelay": 10
+    },
+
+    "VPNMonitorPusherTest": {
```
Comment (ababo) [inline dappctrl-test.config.json]: I don't like "pusher" term in this context. What about "VPNConfigPusher" or "VPNConfigUploader"? Anyway you should use the "config" word in the name.

File: ept/config/client.go [hunk 9]
```diff
@@ -3,109 +3,160 @@ package config
...
+		clientConfName)); err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(filepath.Join(dir, clientAccessName),
+		[]byte(fmt.Sprintf("%s\n%s\n", login, pass)),
+		filePerm)
+}
+
+func autogenFu() string {
```
Comment (ababo) [inline ept/config/client.go]: What is that? Isn't it linked to the tests? If yes, what is a reason to have it here?

File: ept/config/client_test.go [hunk 10]
```diff
@@ -2,36 +2,114 @@ package config
...
+	}
+
+	cliParams, err := parseConfig(confFile,
+		conf.EptTest.ExportConfigKeys, false)
+	if err != nil {
+		t.Fatal(err)
+	}
 
-	conf, err := ClientConfig(address[0], address[1], param)
+	if cfg.Cipher != cliParams[nameCipher] {
```
Comment (ababo) [inline ept/config/client_test.go]: You still didn't get rid of these redundancies.

File: ept/config/server.go [hunk 11]
```diff
@@ -106,7 +106,8 @@ func parseLine(keys map[string]bool,
 	str := strings.TrimSpace(line)
 
 	for key := range keys {
-		if !strings.HasPrefix(str, key) {
+		sStr := strings.Split(str, " ")
+		if sStr[0] != key {
```
Comment (ababo) [inline ept/config/server.go]: Are you confident that sStr cannot be empty?

File: svc/dappvpn/main.go [hunk 12]
```diff
@@ -141,6 +145,31 @@ func handleMonitor() {
...
+			if err := pusher.PushConfig(ctx, c); err != nil {
+				logger.Error("failed to send OpenVpn"+
+					" server configuration: %s\n", err)
+				return
+			}
+
+			conf.Pusher.Pushed = true
+
+			// rewrite dappvpn configuration file
+			if err := writeConfig(confFile, conf); err != nil {
```
Comment (ababo) [inline svc/dappvpn/main.go]: Why do we store OpenVPN configuration in file?

File: ept/config/client.go (hunk 9 above)
Comment (dzeckelev) [reply ept/config/client.go]: this is not related to the tests.
https://golang.org/pkg/text/template/#FuncMap

File: ept/config/client_test.go (hunk 10 above)
Comment (dzeckelev) [reply ept/config/client_test.go]: this is a special test.
The code we discussed earlier is here https://github.com/Privatix/dappctrl/blob/1e3360dfab099897e7a84c4bff1402dde7320bc5/ept/config/client.go#L204

File: ept/config/server.go (hunk 11 above)
Comment (dzeckelev) [reply ept/config/server.go]: Yes it can not be empty.

File: svc/dappvpn/main.go (hunk 12 above)
Comment (dzeckelev) [reply svc/dappvpn/main.go]: We need to overwrite the configuration so that when I next start the program, do not parse the configuration again.

Comment (ababo) [reply svc/dappvpn/main.go]: why not to parse it on every launch? I guess it isn't critical?

Comment (dzeckelev) [reply svc/dappvpn/main.go]: Parsing files is quite a heavy function. Why do it every launch?

Comment (ababo) [reply svc/dappvpn/main.go]: Simpler is better (unless it's a bottleneck).

Comment (ababo) [reply svc/dappvpn/main.go]: By introducing a new config parameter you add additional complexity. As you wish.

**Pull request merged**